### ISSUE TITLE
apache2: multiple Listen directives

### ISSUE THREAD

**Issue by clifforama** [termux/termux-packages#2027]
Having more than one `Listen` directive prevents Apache from serving documents properly -- browser hangs indefinitely when attempting to access any document. Nothing appears in the error log, but an entry does show up in the access log.

This is with Apache 2.4.29 using default httpd.conf. Any ONE of the following `Listen` directives works:

    Listen 8080
    Listen 127.0.0.1:8080
    Listen 127.0.0.2:8080

But any combination of multiple Listen directives fails. Any ideas?

**Comment by fornwall** [comment]
Is this Termux-specific? See https://httpd.apache.org/docs/2.4/bind.html and the `Overlapping Listen directives will result in a fatal error which will prevent the server from starting up` note there.

**Comment by clifforama** [comment]
I think this is Termux-specific, yes. The server does start up without any fatal error. Using just these two non-overlapping `Listen` directives triggers the issue:

    Listen 127.0.0.1:8080
    Listen 127.0.0.2:8080

Can anyone else corroborate this? FWIW, I'm using a Pixel 2 XL on Android 8.1.0.

I've also tried it on Apache 2.4.29 for Windows (via XAMPP) and the same pair of `Listen` directives seems to work.

**Comment by vishalbiswas** [comment]
do separate ports work?

**Comment by vishalbiswas** [comment]
do separate ports work?

**Comment by vishalbiswas** [comment]
do separate ports work?

**Comment by sp1r1t** [comment]
I'm having the same issue. That nasty bug just cost me three hours. Anyway..

If I have just one Listen Directive like
`Listen 8080`
all works fine

When I add a second Listen like
`Listen 8080
Listen 8081`
then Firefox stops getting the Site, responding with "Unable to connect". Similar thing happens for Samsung Internet Browser.

Now interestingly all works fine when in use lynx. The only difference I found was in the apache access logs. Lynx seems to use HTTP 1.0. Whereas the other browsers use HTTP 1.1.

Any ideas on that?

**Comment by terrytech** [comment]
This problem is repeatable for me on my Galaxy S10+ and Moto G7 Edge.  Install termux and apache, then add a second Listen directive and restart apache.  After that, any curl or browser request hangs on requests to either of the two ports I've activated.  

I've confirmed that nginx works fine with similar multi-port http configuration.

Anyone have ideas or clues how to hunt this problem down?

**Comment by dhirenjaypal** [comment]
me too, i cant setup vhost

**Comment by Fernal73** [comment]
Any idea when this will be addressed? Does multiple vmhosts work with one port?

**Comment by xtkoba** [comment]
This should be kept open.

**Comment by xtkoba** [comment]
Here are some observation of `strace` outputs with respect to socket lifecycles.

When I sent

```
GET / HTTP/1.1\r\n
Host: dummy-host.example.com\r\n
\r\n
```

then I got

```
[pid  9262] 16:58:00.034556 accept4(3, {sa_family=AF_INET6, sin6_port=htons(43458), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "::ffff:192.168.100.121", &sin6_addr), sin6_scope_id=0}, [128 => 28], 0) = 15
[pid  9262] 16:58:00.046594 fcntl(15, F_GETFD) = 0
[pid  9262] 16:58:00.048302 fcntl(15, F_SETFD, FD_CLOEXEC) = 0
[pid  9193] 16:58:00.054832 getsockname(15, {sa_family=AF_INET6, sin6_port=htons(8080), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "::ffff:192.168.100.111", &sin6_addr), sin6_scope_id=0}, [128 => 28]) = 0
[pid  9193] 16:58:00.056546 read(15, "GET / HTTP/1.1\r\nHost: dummy-host"..., 8000) = 48
[pid  9193] 16:58:00.111376 read(15,  <unfinished ...>
```

and the server continued to read on the socket indefinitely, with no response, which is unexpected.

On the other hand, when I sent

```
GET / HTTP/1.0\r\n
\r\n
```

then I got

```
[pid  9411] 17:01:00.824512 accept4(3, {sa_family=AF_INET6, sin6_port=htons(43462), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "::ffff:192.168.100.121", &sin6_addr), sin6_scope_id=0}, [128 => 28], 0) = 15
[pid  9411] 17:01:00.838266 fcntl(15, F_GETFD) = 0
[pid  9411] 17:01:00.839457 fcntl(15, F_SETFD, FD_CLOEXEC) = 0
[pid  9343] 17:01:00.844886 getsockname(15, {sa_family=AF_INET6, sin6_port=htons(8080), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "::ffff:192.168.100.111", &sin6_addr), sin6_scope_id=0}, [128 => 28]) = 0
[pid  9343] 17:01:00.846614 read(15, "GET / HTTP/1.0\r\n\r\n", 8000) = 18
[pid  9343] 17:01:00.895785 writev(15, [{iov_base="HTTP/1.1 200 OK\r\nDate: Wed, 01 D"..., iov_len=170}, {iov_base="<!DOCTYPE HTML PUBLIC \"-//W3C//D"..., iov_len=161}], 2) = 331
[pid  9343] 17:01:00.930320 shutdown(15, SHUT_WR) = 0
[pid  9343] 17:01:00.937246 read(15, "", 512) = 0
[pid  9343] 17:01:00.938558 close(15)   = 0
```

which is expected.

**Comment by xtkoba** [comment]
With only one Listen directive specified, I had

```
[pid  7077] 05:35:51.769516 accept4(3,  <unfinished ...>
[pid  7077] 05:36:00.308653 <... accept4 resumed>{sa_family=AF_INET6, sin6_port=htons(47996), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "::ffff:192.168.100.121", &sin6_addr), sin6_scope_id=0}, [128 => 28], 0) = 12
[pid  7077] 05:36:00.321174 fcntl(12, F_GETFD) = 0
[pid  7077] 05:36:00.322522 fcntl(12, F_SETFD, FD_CLOEXEC) = 0
[pid  7011] 05:36:00.327435 getsockname(12, {sa_family=AF_INET6, sin6_port=htons(8080), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "::ffff:192.168.100.111", &sin6_addr), sin6_scope_id=0}, [128 => 28]) = 0
[pid  7011] 05:36:00.330018 fcntl(12, F_GETFL) = 0x2 (flags O_RDWR)
[pid  7011] 05:36:00.332353 fcntl(12, F_SETFL, O_RDWR|O_NONBLOCK) = 0
[pid  7011] 05:36:00.334953 read(12, "GET / HTTP/1.1\r\nHost: dummy-host"..., 8000) = 48
[pid  7011] 05:36:00.411321 read(12, 0x7d75d08048, 8000) = -1 EAGAIN (Try again)
[pid  7011] 05:36:00.414059 writev(12, [{iov_base="HTTP/1.1 200 OK\r\nDate: Wed, 01 D"..., iov_len=151}, {iov_base="<!DOCTYPE HTML PUBLIC \"-//W3C//D"..., iov_len=161}], 2) = 312
```

This time the second `read(2)` (timestamped at 05:36:00.411321) returns immediately with `EAGAIN` error, because the file descriptor is flagged as `O_NONBLOCK`.

Now I remember that on Android `O_NONBLOCK` flag is not inherited to a child process (see #7723), which seems to be the culprit here.

I have not yet come up with how to fix this, but it can be said that as of now the `apr` package (which `apache2` depends on) is misconfigured so that it incorrectly assumes `O_NONBLOCK` is inherited.

**Comment by xtkoba** [comment]
Just replacing `lib/libapr-1.so` with the one from https://github.com/termux/termux-packages/actions/runs/1328779993 seems to fix this.

So a possible fix is to merge #7723. While this specific PR is open (and not rejected explicitly) I'm not going to open another PR, but anyone can do that.
